Source/Dafny/Compiler-js.cs

```diff
@@ -1042,9 +1042,51 @@ protected class ClassWriter : IClassWriter {
     // ----- Statements -------------------------------------------------------------
 
     protected override void EmitPrintStmt(TargetWriter wr, Expression arg) {
-      wr.Write("process.stdout.write(_dafny.toString(");
-      TrExpr(arg, wr, false);
-      wr.WriteLine("));");
+      bool isString = arg.Type.AsCollectionType != null &&
+                      arg.Type.AsCollectionType.AsSeqType != null &&
+                      arg.Type.AsCollectionType.AsSeqType.Arg.IsCharType;
+      bool isStringLiteral = arg is StringLiteralExpr;
+      bool isGeneric = arg.Type.AsCollectionType != null &&
+                       arg.Type.AsCollectionType.AsSeqType != null &&
+                       arg.Type.AsCollectionType.AsSeqType.Arg.IsTypeParameter;
+      if (isString && isStringLiteral) {
+        // process.stdout.write(_dafny.toString(x));
+        wr.Write("process.stdout.write(_dafny.toString(");
+        TrExpr(arg, wr, false);
+        wr.WriteLine("));");
+      } else if (isString && !isStringLiteral) {
+        // try { process.stdout.write(_dafny.toString((x).join(""))); } catch (_error) { process.stdout.write(_dafny.toString(x)); }
+        wr.Write("try { process.stdout.write(_dafny.toString((");
+        TrExpr(arg, wr, false);
+        wr.WriteLine(").join(\"\")));");
+        wr.Write(" } catch (_error) { process.stdout.write(_dafny.toString(");
+        TrExpr(arg, wr, false);
+        wr.WriteLine(")); }");
+      } else if (!isString && !isGeneric) {
```

Comment:
With L1052 and L 1057 simplified, this guard can be just `!isGeneric`

Reply:
Reordered the cases so that the guards are simplified without losing clarity